client/cloudevents/sourceclient.go

```diff
@@ -0,0 +1,315 @@
+package cloudevents
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/google/uuid"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	"open-cluster-management.io/api/client/cloudevents/payload"
+	"open-cluster-management.io/api/client/cloudevents/protocol"
+	"open-cluster-management.io/api/client/cloudevents/types"
+)
+
+type CloudEventSourceClient[T ResourceObject] struct {
+	contextGetter    protocol.CloudEventsContextGetter
+	subClient        cloudevents.Client
+	pubClient        cloudevents.Client
+	lister           Lister[T]
+	codecs           map[types.GroupVersionResource]Codec[T]
+	statusHashGetter StatusHashGetter[T]
+	sourceID         string
+}
+
+func NewCloudEventSourceClient[T ResourceObject](
```

Comment:
comments on all input var

Reply:
added

Reply:
there seems to be something missing, e.g how to setup the sub/pub client for agent and source client, will it be same context or they need to use the different context, who takes the responsibility to set the context?

Reply:
I try to introduce a `CloudEventsClient` interface to avoid using the cloudevents clients and context directly, please take a look